### PR TITLE
Cale/add baseten models

### PR DESCRIPTION
**Pull request by coval-cale**
## Summary by CodeRabbit

* **New Features**
  * Added a benchmark probe command that runs selected STT/TTS models and returns aggregated JSON results.
  * Added an `stt-smoke` command that validates WAV format, streams speech-to-text, and outputs JSON timing/transcript results.
  * Added Baseten-backed streaming STT/TTS provider support and registered Baseten STT/TTS models.

* **Bug Fixes**
  * Updated `tts-smoke` to enforce a 120s timeout and emit a failure JSON event with a non-zero exit code.

* **Tests**
  * Added mocked async test coverage for Baseten STT/TTS and for probe aggregation/non-persist execution.

<h3>Greptile Summary</h3>

This PR adds Baseten dedicated-endpoint providers for STT (Whisper Large V3 over WebSocket) and TTS (Qwen3-TTS via WebSocket), registers them as `PENDING` models, wires them into the orchestrator, and adds a new `probe` CLI command for non-persisting latency checks plus an `stt-smoke` command and a timeout guard on `tts-smoke`.

- **Baseten STT/TTS providers** (`providers/stt/baseten.py`, `providers/tts/baseten.py`): both implement the respective WebSocket streaming protocols with proper error handling, auth headers, and cold-start open-timeout settings; a pre-existing flagged issue with `audio_start_time` being overwritten on every sent chunk (rather than set once at stream start) affects TTFT and `audio_to_final` accuracy.
- **Probe runner** (`runner/probe.py`): non-persisting benchmark runner that reuses `_run_stt_item`/`_run_tts_item` with `writer=None`, aggregates metrics (mean/median/p90/min/max), and is exposed via a new `probe` CLI command.
- **CLI additions** (`__main__.py`): `stt-smoke` validates WAV format before streaming, both smoke commands gain 120 s timeouts with structured JSON failure output; provider-init errors (e.g. missing env vars) still surface as raw tracebacks rather than clean exit-code-1 JSON.

<h3>Confidence Score: 4/5</h3>

Safe to merge for PENDING/internal use; the existing unresolved bug in `_send_audio` causes TTFT and audio-to-final measurements to be anchored to the last sent audio frame rather than the first, so the STT metrics collected during probe runs will be systematically inaccurate until fixed.

The STT provider's `audio_start_time` is overwritten on every chunk, meaning all latency measurements (`ttft_seconds`, `audio_to_final_seconds`) are computed relative to the most recently sent frame rather than the beginning of the audio stream. Since these metrics drive the benchmark's TTFS computation in the orchestrator, any probe or production run with this provider will produce misleadingly low latency numbers.

runner/src/coval_bench/providers/stt/baseten.py — specifically `_send_audio` and the `audio_start_time` assignment inside the per-chunk loop.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| runner/src/coval_bench/providers/stt/baseten.py | New Baseten STT WebSocket provider; contains the flagged P1 where `audio_start_time` is overwritten on every chunk instead of being set once at stream start, causing systematically wrong TTFT and audio_to_final metrics. |
| runner/src/coval_bench/providers/tts/baseten.py | New Baseten TTS WebSocket provider; synthesize logic and error handling look solid; `name` property returns a compound string inconsistent with every other provider (flagged separately). |
| runner/src/coval_bench/runner/probe.py | New non-persisting probe runner; aggregation logic and result-key scheme are clean for the single-voice-per-model registry entries added here; all items are gathered concurrently with semaphore control. |
| runner/src/coval_bench/__main__.py | Adds `stt-smoke`, `probe`, and timeout handling for `tts-smoke`; provider-init errors in both smoke commands still surface as raw tracebacks (flagged separately); probe command has no top-level timeout, relying on per-item orchestrator timeouts. |
| runner/src/coval_bench/runner/orchestrator.py | Adds `baseten` branch to the STT provider kwargs in `_run_stt_item`; minimal and consistent with the existing `google` branch pattern. |
| runner/src/coval_bench/config.py | Adds `baseten_api_key`, `baseten_whisper_url`, and `baseten_qwen_url` to Settings; all nullable to maintain backward compatibility. |
| runner/src/coval_bench/registries/models.py | Registers Baseten STT (whisper-large-v3) and TTS (qwen3-tts-1.7b) models as PENDING with DEDICATED tenancy; correct attributes and consistent with other registry entries. |
| runner/tests/providers/stt/test_baseten.py | Good test coverage for STT provider: happy path, multi-segment finals, acknowledged vs. finished end_audio, error event, wrong sample rate, connection failure, and stream-ends-without-final. |
| runner/tests/providers/tts/test_baseten.py | Comprehensive TTS test suite; validates wire protocol order, TTFA timing, auth headers, error events, and init validation guards. |
| runner/tests/runner/test_probe.py | Probe tests verify `writer=None` propagation and `_aggregate` correctness including edge case of all-failed rows; uses monkeypatching so no live I/O. |

<h3>Sequence Diagram</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant CLI as CLI (stt-smoke / probe)
    participant Provider as BasetenSTTProvider
    participant WS as Baseten WebSocket
    participant Send as _send_audio task
    participant Recv as _receive task

    CLI->>Provider: measure_ttft(pcm, 1, 2, 16000)
    Provider->>WS: connect(ws_url, Authorization header)
    Provider->>WS: send streaming_params config (JSON)
    Provider->>Send: asyncio.create_task(_send_audio)
    Provider->>Recv: asyncio.create_task(_receive)
    Provider->>Provider: asyncio.wait(FIRST_EXCEPTION)

    loop paced 512-sample frames
        Send->>Send: "result.audio_start_time = frame_start (overwritten each frame)"
        Send->>WS: send binary PCM frame
    end
    Send->>WS: send end_audio JSON

    WS-->>Recv: "transcription is_final=false"
    Recv->>Recv: "ttft = now - result.audio_start_time"
    WS-->>Recv: "end_audio status=acknowledged"
    Note over Recv: continue (not finished yet)
    WS-->>Recv: "transcription is_final=true"
    Recv->>Recv: "audio_to_final = now - result.audio_start_time"
    WS-->>Recv: "end_audio status=finished"
    Recv->>Recv: break loop

    Provider->>Provider: asyncio.gather(send_task, recv_task)
    Provider-->>CLI: TranscriptionResult(ttft, audio_to_final, transcript)
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant CLI as CLI (stt-smoke / probe)
    participant Provider as BasetenSTTProvider
    participant WS as Baseten WebSocket
    participant Send as _send_audio task
    participant Recv as _receive task

    CLI->>Provider: measure_ttft(pcm, 1, 2, 16000)
    Provider->>WS: connect(ws_url, Authorization header)
    Provider->>WS: send streaming_params config (JSON)
    Provider->>Send: asyncio.create_task(_send_audio)
    Provider->>Recv: asyncio.create_task(_receive)
    Provider->>Provider: asyncio.wait(FIRST_EXCEPTION)

    loop paced 512-sample frames
        Send->>Send: "result.audio_start_time = frame_start (overwritten each frame)"
        Send->>WS: send binary PCM frame
    end
    Send->>WS: send end_audio JSON

    WS-->>Recv: "transcription is_final=false"
    Recv->>Recv: "ttft = now - result.audio_start_time"
    WS-->>Recv: "end_audio status=acknowledged"
    Note over Recv: continue (not finished yet)
    WS-->>Recv: "transcription is_final=true"
    Recv->>Recv: "audio_to_final = now - result.audio_start_time"
    WS-->>Recv: "end_audio status=finished"
    Recv->>Recv: break loop

    Provider->>Provider: asyncio.gather(send_task, recv_task)
    Provider-->>CLI: TranscriptionResult(ttft, audio_to_final, transcript)
```

</a>

<details><summary><h3>Comments Outside Diff (3)</h3></summary>

1. `runner/src/coval_bench/__main__.py`, line 93 ([link](https://github.com/coval-ai/benchmarks/blob/6518f35d9e0d0dc337029fbfc72e7360a4f3ec18/runner/src/coval_bench/__main__.py#L93)) 

   <a href="#">P2</a> **Provider init errors surface as unhandled tracebacks**

   The docstring contracts exit code 1 for provider errors and exit code 2 for unknown providers, but `provider_cls(**kwargs)` is not wrapped in a try/except. If `baseten_whisper_url` or `baseten_api_key` is absent from the environment, `BasetenSTTProvider.__init__` raises `ValueError`, which propagates as an uncaught exception (messy traceback) rather than a clean `click.echo(..., err=True)` + `sys.exit(1)` sequence.

2. `runner/src/coval_bench/providers/tts/baseten.py`, line 481-482 ([link](https://github.com/coval-ai/benchmarks/blob/6518f35d9e0d0dc337029fbfc72e7360a4f3ec18/runner/src/coval_bench/providers/tts/baseten.py#L481-L482)) 

   <a href="#">P2</a> **`name` property returns a compound string unlike every other provider**

   The STT counterpart returns `"baseten"`, every other provider in the codebase returns a flat provider name, but this property returns `"baseten-qwen3-tts-1.7b"`. The `finalize_tts_result` call hardcodes `provider="baseten"` so the DB record is unaffected, but any logging or display code that reads `provider.name` will see an inconsistent value. Consider aligning with the STT provider and returning `"baseten"`.

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!

3. `runner/src/coval_bench/providers/stt/baseten.py`, line 367-374 ([link](https://github.com/coval-ai/benchmarks/blob/36d8b8147737fd233a530a25be1db2373baf2d9a/runner/src/coval_bench/providers/stt/baseten.py#L367-L374)) 

   <a href="#">P1</a> **`audio_start_time` overwritten on every chunk — TTFT and `audio_to_final` measured from last sent frame, not from audio start**

   `result.audio_start_time` is reassigned on each loop iteration in `_send_audio`, so by the time a transcription message arrives in `_receive`, it holds the timestamp of the most recently sent 512-sample frame rather than the timestamp of the first frame. This means `ttft_seconds` measures "latency from last audio frame before first token" (roughly server processing latency per frame) and `audio_to_final_seconds` measures "time from last sent chunk to final transcript." The latter is used by the orchestrator to compute TTFS via `compute_ttfs(audio_to_final, speech_end_offset_ms / 1000.0)`, so if `audio_to_final` is not anchored to the start of the audio stream (as it is for every other STT provider), the TTFS value will be systematically lower than the true figure and incomparable to the rest of the cohort.
</details>

<sub>Reviews (5): Last reviewed commit: ["Require non-empty transcript for stt-smo..."](https://github.com/coval-ai/benchmarks/commit/6ce3ef33c82c4d3b9f5c1d427120362b1780dea4) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=40973198)</sub>

### CODE REVIEW

File: runner/src/coval_bench/providers/stt/baseten.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        try:
+            async for chunk, start in paced_chunks(audio_data, _FRAME_BYTES, _BYTE_RATE):
+                # Stop early if _receive already recorded a protocol/auth error.
+                if result.error is not None:
+                    break
+                result.audio_start_time = start
+                frame = chunk
+                if len(frame) < _FRAME_BYTES:  # pad the final short frame to 512 samples
+                    frame += b"\x00" * (_FRAME_BYTES - len(frame))
+                await ws.send(frame)
```
[the remark this answers is not in the document]
Comment (coval-cale) [reply runner/src/coval_bench/providers/stt/baseten.py:151]: wrong. `start` is not a per frame stamp.

File: runner/src/coval_bench/providers/tts/baseten.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                    if isinstance(message, (bytes, bytearray)):
+                        if first_chunk_at is None:
+                            first_chunk_at = time.monotonic()
+                        audio_chunks.append(bytes(message))
+                        continue
+                    data: dict[str, Any] = json.loads(message)
+                    if data.get("type") == "session.done":
+                        break
+                    if data.get("type") == "error":
+                        raise RuntimeError(str(data.get("message", data)))
```
[the remark this answers is not in the document]
Comment (coval-cale) [reply runner/src/coval_bench/providers/tts/baseten.py:106]: Fixed

File: runner/src/coval_bench/providers/stt/baseten.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        result: TranscriptionResult,
+    ) -> None:
+        try:
+            async for chunk, start in paced_chunks(audio_data, _FRAME_BYTES, _BYTE_RATE):
+                # Stop early if _receive already recorded a protocol/auth error.
+                if result.error is not None:
+                    break
+                result.audio_start_time = start
+                frame = chunk
+                if len(frame) < _FRAME_BYTES:  # pad the final short frame to 512 samples
```
Comment (seribaymadina) [inline runner/src/coval_bench/providers/stt/baseten.py:149]: not sure how important is this, but seems like baseten's protocol may impose up to one  VAD frame of extra tailing silence. can we confirm/reject this?

Comment (coval-cale) [reply runner/src/coval_bench/providers/stt/baseten.py:149]: Nice find. Thankfully it doesn't affect the value of `end_audio` so we're all good.

File: runner/src/coval_bench/providers/stt/baseten.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+            # Signal end of audio so the server flushes the final transcript.
+            await ws.send(json.dumps({"type": "end_audio"}))
+        except Exception as exc:
+            logger.warning(
+                "baseten_send_error", provider="baseten", model=self._model, exc_info=exc
+            )
+            raise
+
+    async def _receive(self, ws: Any, result: TranscriptionResult) -> None:
+        # Per-segment finals accumulate in arrival order (Whisper + Silero VAD).
```
Comment (seribaymadina) [inline runner/src/coval_bench/providers/stt/baseten.py:161]: interesting that they are internally using Silero VAD for the endpointing behavior.  Wondering if for other inference providers this could be different? and if so, considering our dataset anchoring on end of speech based on Silero VAD as well for TTFS measurement , could that be a factor for us to switch to another external VAD

**Review by seribaymadina — APPROVED** [review]
super awesome job done on this! left two comments/ questions

File: runner/src/coval_bench/providers/stt/baseten.py (hunk 4 above)
Comment (coval-cale) [reply runner/src/coval_bench/providers/stt/baseten.py:161]: I dig this.
